### PR TITLE
Compile guide for Trainer

### LINKED ISSUE #18852: Add `torch.compile` guide to docs
### 📚 Documentation

This applies to 

- [x] Fabric (https://github.com/Lightning-AI/pytorch-lightning/pull/19330)
- [ ] PyTorch Lightning

Things to mention:
- Compiling `nn.Module` vs `LightningModule` or `fabric.setup(module)` (#17250)
- Known limitations.
- Known sources of graph breaks.

Related: #12398

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds a similar guide to PyTorch Lightning as done in #19330 for Fabric.

Fixes #18852

----
📚 Documentation preview 📚: https://pytorch-lightning--19531.org.readthedocs.build/en/19531/

cc @borda @carmocca

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/advanced/compile.rst [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+----
+
+
+**************************************
+A note about torch.compile in practice
+**************************************
+
+In practice, you will find that ``torch.compile`` often doesn't work well and can even be counter-productive.
+Compilation may fail with cryptic error messages that are impossible to debug without help from the PyTorch team.
```
Comment (lantiga) [inline docs/source-pytorch/advanced/compile.rst:240]:
```suggestion
Compilation may fail with cryptic error messages that are hard to debug, luckily the PyTorch team is responsive and it's likely that messaging will improve in time.
```

File: docs/source-pytorch/advanced/compile.rst [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+It is also not uncommon that ``torch.compile`` will produce a significantly *slower* model or one with much higher memory usage.
+On top of that, the compilation phase itself can be incredibly slow, taking several minutes to finish.
```
Comment (lantiga) [inline docs/source-pytorch/advanced/compile.rst:242]:
```suggestion
As a note, the compilation phase itself will take some time, taking up to several minutes.
```

File: docs/source-pytorch/advanced/compile.rst [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+For these reasons, we recommend that you don't waste too much time trying to apply ``torch.compile`` during development, and rather evaluate its effectiveness toward the end when you are about to launch long-running, expensive experiments.
```
Comment (lantiga) [inline docs/source-pytorch/advanced/compile.rst:243]:
```suggestion
For these reasons, we recommend that you don't invest too much time trying to apply ``torch.compile`` during development, and rather evaluate its effectiveness toward the end when you are about to launch long-running, expensive experiments.
```

**Review by lantiga — APPROVED** [review]
Looks great, just a few changes

**Pull request merged**